### PR TITLE
http conn man: add tracing config for path length in tag

### PR DESCRIPTION
**Pull request by douglas-reid**
Description:

This PR adds a configuration option for controlling the length of the request path that is included in the HttpUrl span tag. Currently, this length is hard-coded to 256. With this PR, that length will be configurable (defaulting to the old value).

Risk Level: Low

Testing: Unit

Docs Changes: Inline with the API proto. Documented new field.

Release Notes: Added in the PR.

Related issue: https://github.com/istio/istio/issues/14563

cc: @Stono @objectiser

### CODE REVIEW

File: test/common/tracing/http_tracer_impl_test.cc [hunk 1]
```diff
@@ -336,6 +336,7 @@ TEST(HttpConnManFinalizerImpl, SpanPopulatedFailureResponse) {
   EXPECT_CALL(span, setTag(Eq("cc"), Eq("c")));
   EXPECT_CALL(config, requestHeadersForTags());
   EXPECT_CALL(config, verbose).WillOnce(Return(false));
+  EXPECT_CALL(config, maxPathTagLength).WillOnce(Return(256));
```
Comment (objectiser) [inline test/common/tracing/http_tracer_impl_test.cc]: Would it be worth setting the length to something lower and check that the `http.url` tag actually has a truncated value?

Comment (objectiser) [reply test/common/tracing/http_tracer_impl_test.cc]: Yeah that is fine.

**Comment by objectiser** [comment]
Sounds reasonable to add to docs. LTGM.

**Comment by Stono** [comment]
🎉

File: api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto [hunk 2]
```diff
@@ -127,6 +127,11 @@ message HttpConnectionManager {
     // Whether to annotate spans with additional data. If true, spans will include logs for stream
     // events.
     bool verbose = 6;
+
+    // Maximum length of the request path to extract and include in the HttpUrl tag. Used to
+    // truncate lengthy request paths to meet the needs of a tracing backend.
+    // Default: 256
+    uint32 max_path_tag_length = 7;
```
Comment (htuch) [inline api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto]: Please use `google.protobuf.UInt32Value` for defaulted integers.

File: docs/root/intro/version_history.rst [hunk 3]
```diff
@@ -52,6 +52,7 @@ Version history
   certificate validation context.
 * tracing: added tags for gRPC response status and meesage.
 * upstream: added :ref:`an option <envoy_api_field_Cluster.CommonLbConfig.close_connections_on_host_set_change>` that allows draining HTTP, TCP connection pools on cluster membership change.
+* tracing: added :ref:`max_path_tag_length <envoy_api_field_config.filter.network.http_connection_manager.v2.HttpConnectionManager.tracing>` to support customizing the length of the request path included in the extracted HttpUrl tag.
```
Comment (htuch) [inline docs/root/intro/version_history.rst]: Nit: you might want to do a `:ref:` back to `HttpUrl` message definition as well.

File: source/common/tracing/http_tracer_impl.h [hunk 4]
```diff
@@ -118,6 +118,7 @@ class EgressConfigImpl : public Config {
     return request_headers_for_tags_;
   }
   bool verbose() const override { return false; }
+  uint32_t maxPathTagLength() const override { return 256; }
```
Comment (htuch) [inline source/common/tracing/http_tracer_impl.h]: Can we avoid hardcoding 256 in places? Suggest a shared constant.

File: api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto (hunk 2 above)
Comment (douglas-reid) [reply api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto]: done.

File: docs/root/intro/version_history.rst (hunk 3 above)
Comment (douglas-reid) [reply docs/root/intro/version_history.rst]: added a pointer to the external definition of the tag. that ok?

File: source/common/tracing/http_tracer_impl.h (hunk 4 above)
Comment (douglas-reid) [reply source/common/tracing/http_tracer_impl.h]: wasn't quite sure where to put it, think I found a reasonable spot. updated useful places to use that constant.

**Comment by douglas-reid** [comment]
Updated, attempting to resolve all raised issues. PTAL.

File: source/extensions/filters/network/http_connection_manager/config.cc [hunk 5]
```diff
@@ -274,10 +275,14 @@ HttpConnectionManagerConfig::HttpConnectionManagerConfig(
     overall_sampling.set_numerator(
         tracing_config.has_overall_sampling() ? tracing_config.overall_sampling().value() : 100);
 
+    uint32_t max_path_tag_length = tracing_config.has_max_path_tag_length()
```
Comment (htuch) [inline source/extensions/filters/network/http_connection_manager/config.cc]: Nit: suggest making this `const` and using https://github.com/envoyproxy/envoy/blob/master/source/common/protobuf/utility.h#L20.

Comment (douglas-reid) [reply source/extensions/filters/network/http_connection_manager/config.cc]: Done.

**Comment by douglas-reid** [comment]
Addressed the nit. Thanks for the pointer. PTAL.

**Comment by douglas-reid** [comment]
Looks like tsan failed to a time out in an integration test?

**Pull request merged**